src/App.tsx

```diff
@@ -1,26 +1,184 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
-import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import React, { useEffect, useMemo, useState } from 'react';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+import { addTodo, deleteTodo, getTodos, updateTodo } from './api/todos';
+import { Filters, LoadingType } from './types';
+import { TodoHeader } from './components/TodoHeader';
+import { TodoFooter } from './components/TodoFooter';
+import { ErrorNotification } from './components/ErrorNotification';
+import { makeLoadingObject } from './utils/makeLoadingObject';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState<Filters>(Filters.All);
+  const [loadingIds, setLoadingIds] = useState<LoadingType>({});
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+        // throw new Error('Unable to load todos');
+      });
+  }, []);
+
+  const handleAddTodo = (newTodo: Todo): Promise<Todo | void> => {
+    setTempTodo(newTodo);
+
+    return addTodo(newTodo).then(newTodoRes => {
+      setTodos(prevTodos => [...prevTodos, newTodoRes]);
+    });
+  };
+
+  const updateCompletedTodo = (
+    updTodo: Todo,
+    key: keyof Todo,
+    value: boolean | string,
+  ) => {
+    return updateTodo({ ...updTodo, [key]: value })
+      .then((respTodo: Todo) => {
+        setTodos(currTodos => {
+          return currTodos.map(todo =>
+            todo.id === updTodo.id ? respTodo : todo,
+          );
+        });
+
+        return false;
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+
+        return true;
+      });
+  };
+
+  const handleToggleTodos = () => {
+    const isActive = todos.filter(todo => !todo.completed);
+    const isActiveIds = makeLoadingObject(isActive);
+
+    if (isActive.length > 1) {
+      setLoadingIds(isActiveIds);
+
+      Promise.all(
+        isActive.map(todo => updateTodo({ ...todo, completed: true })),
+      )
+        .then(() =>
+          setTodos(prevTodos => {
+            return prevTodos.map(todo2 => {
```

Comment:
```suggestion
            return previous.map(todo => {
```
Please try to give clear names